pkg/gather/gather.go

```diff
@@ -0,0 +1,254 @@
+	archiveMetadataFilename         = "insights-operator/archive-metadata.json"
+	allGatherersConst               = "ALL"
+	// exists in every archive
+	clusterConfigGathererPeriodMultiplier = 1
+	// exists in half of the archives
+	someGathererPeriodMultiplier = 2
+)
+
+var programStartTime time.Time
+
+func init() {
+	programStartTime = time.Now()
+}
+
+// GathererFunctionReport contains the information about a specific gathering function
+type GathererFunctionReport struct {
+	FuncName     string        `json:"name"`
+	Duration     time.Duration `json:"duration_in_ms"`
+	RecordsCount int           `json:"records_count"`
+	Errors       []error       `json:"errors"`
+	Panic        interface{}   `json:"panic"`
+}
+
+// GathererStatusReport contains the information about the data in the archive related to a specific gatherer
+type GathererStatusReport struct {
+	// info about gathering functions
+	StatusReports []GathererFunctionReport `json:"status_reports"`
+}
+
+// ArchiveMetadata contains the information about the archive regardless of what gatherers it contains
+type ArchiveMetadata struct {
+	// MemoryAlloc is the amount of memory taken by heap objects after processing the records
+	MemoryAlloc uint64 `json:"memory_alloc_bytes"`
+	// Uptime is the number of seconds from the program start till the point when metadata was created
+	Uptime float64 `json:"uptime_seconds"`
+	// IsGlobalObfuscationEnabled shows if obfuscation(hiding IPs and cluster domain) is enabled
+	IsGlobalObfuscationEnabled bool `json:"is_global_obfuscation_enabled"`
+}
+
+// CreateAllGatherers creates all the gatherers with their periods
+func CreateAllGatherers(
+	gatherKubeConfig, gatherProtoKubeConfig, metricsGatherKubeConfig *rest.Config,
+	anonymizer *anonymization.Anonymizer,
+) map[uint]Interface {
+	clusterConfigGatherer := clusterconfig.New(
+		gatherKubeConfig, gatherProtoKubeConfig, metricsGatherKubeConfig, anonymizer,
+	)
+	someGatherer := some_gatherer.New()
+
+	return map[uint]Interface{
+		clusterConfigGathererPeriodMultiplier: clusterConfigGatherer,
+		someGathererPeriodMultiplier:          someGatherer,
+	}
+}
+
+// CollectAndRecord gathers enabled functions of the provided gatherer and records the results to the recorder
+func CollectAndRecord(
+	ctx context.Context,
+	gatherer Interface,
+	rec recorder.Interface,
+	configurator configobserver.Configurator,
+	anonymizer *anonymization.Anonymizer,
+) error {
+	resultsChan, err := StartGatheringConcurrently(ctx, gatherer, configurator.Config().Gather)
+	if err != nil {
+		return err
+	}
+
+	gathererName := gatherer.GetName()
+
+	var errs []error
+	var functionReports []GathererFunctionReport
+
+	for result := range resultsChan {
+		if result.Panic != nil {
+			errStr := fmt.Sprintf(
+				"gatherer %v's function %v panicked with error: %v",
+				gathererName, result.FunctionName, result.Panic,
+			)
+
+			if result.IgnoreErrors {
+				klog.Error(errStr)
+			} else {
+				panic(errStr)
```

Comment:
Checking panic of particular gatherers is a good idea IMO, but do we want to do this? It looks like that nothing will be gathered for the given gatherer and there isn't any related info in the metadata.

Reply:
There is, I added panic field to metadata as well. 

https://github.com/openshift/insights-operator/pull/399/files#diff-bdd45c95687562d23e5fdd43dea644d8fb86c95354192cedd72835743ba9fc0fR45

The idea of this whole catching panic thing is to make the IO more stable for failable gatherers.

Reply:
Then `panic(errStr)` will return panic back and it will be the same as existing behaviour.

Reply:
I don't think so. Right now when there's a panic -> container terminates AFAIK -> pod restart and when experimenting with this draft  I have seen the behavior mentioned above...maybe I am missing something or was doing something wrong.

Reply:
@tremes it's actually previous code not working correctly. I found out that exponential backoff which we use here https://github.com/openshift/insights-operator/blob/master/pkg/controller/periodic/periodic.go#L104 panics only when `ReallyCrash` set to `true` https://github.com/openshift/insights-operator/blob/master/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go#L53 (what an unobvious way...) which is set in this file https://github.com/openshift/insights-operator/blob/master/vendor/github.com/openshift/library-go/pkg/serviceability/panic.go by `BehaviorOnPanic` which is called in some places in the libs(`grep -r BehaviorOnPanic`), but also in our code here https://github.com/openshift/insights-operator/blob/master/pkg/cmd/start/start.go#L123 as you can see from the file https://github.com/openshift/insights-operator/blob/master/vendor/github.com/openshift/library-go/pkg/serviceability/panic.go there are multiple options and if you set `OPENSHIFT_ON_PANIC` env variable to `crash` it's gonna panic, if you keep it empty, the default behaviour of just logging an error is used. Why did the previous code stop? Because the panic happened in the goroutine and couldn't be catched by exponential backoff.

Reply:
Good findings. I am not saying that the original behavior is correct. I am just saying that right now (with your PR) we will not fail (container terminates), but it may also happen that there will be almost empty archive with no corresponding information in the metadata. 

In other words this is panicking in the main goroutine (if I am not mistaken) and that is not necessary IMO.

Reply:
@tremes, I've changed the logic so that now panics are logged and treated as errors. In case there was a panic in a gathering function we won't crash the operator and get nice and healthy archive with all other collected data and panic information about this bad function, which in my opinion is better than crashing cuz in case of a crash we won't get any archive. WDYT?

Reply:
Sorry I missed this one. Yeah sounds good.